### PR TITLE
Fix broken SET reuse during projection analysis.

### LINKED ISSUE #35336: Result of a column using `if` and a subquery changing when using `WHERE 1=1` as a filter
**Describe what's wrong**

When building a query using `if` and a subquery as a column, the inclusion of a trivial `WHERE` modifies the column.

**Does it reproduce on recent release?**

Yes, reproduced with 22.2.3.5.

**How to reproduce**

1. `docker run --rm -it clickhouse/clickhouse-server:22.2.3.5`
2. `clickhouse client`

```sql
CREATE TABLE IF NOT EXISTS flows
(
    Date Date,
    TimeReceived DateTime,
    SrcAS UInt32,
    Bytes UInt64
) ENGINE = MergeTree()
PARTITION BY Date
ORDER BY TimeReceived;

INSERT INTO TABLE flows VALUES
(toDate('2022-03-16 15:04:00'), toDateTime('2022-03-16 15:04:00'), 15169, 83948),
(toDate('2022-03-16 15:04:00'), toDateTime('2022-03-16 15:04:00'), 12322, 98989),
(toDate('2022-03-16 15:04:00'), toDateTime('2022-03-16 15:04:00'), 60068, 99990),
(toDate('2022-03-16 15:04:00'), toDateTime('2022-03-16 15:04:00'), 15169, 89898),
(toDate('2022-03-16 15:05:00'), toDateTime('2022-03-16 15:05:00'), 15169, 83948),
(toDate('2022-03-16 15:05:00'), toDateTime('2022-03-16 15:05:00'), 15169, 89898),
(toDate('2022-03-16 15:06:00'), toDateTime('2022-03-16 15:06:00'), 15169, 83948),
(toDate('2022-03-16 15:07:00'), toDateTime('2022-03-16 15:07:00'), 15169, 89898),
(toDate('2022-03-16 15:08:00'), toDateTime('2022-03-16 15:08:00'), 15169, 83948),
(toDate('2022-03-16 15:09:00'), toDateTime('2022-03-16 15:09:00'), 15169, 89898),
(toDate('2022-03-16 15:10:00'), toDateTime('2022-03-16 15:10:00'), 15169, 83948),
(toDate('2022-03-16 15:11:00'), toDateTime('2022-03-16 15:11:00'), 15169, 89898);

SELECT toStartOfInterval(toDateTime(TimeReceived), INTERVAL 20 second) AS TimeReceived,
       if(SrcAS in (SELECT SrcAS FROM flows GROUP BY SrcAS ORDER BY SUM(Bytes) DESC LIMIT 10) , SrcAS, 33) AS SrcAS
FROM flows
WHERE 2 == 2
```

This yields:

```
┌────────TimeReceived─┬─SrcAS─┐
│ 2022-03-16 15:04:00 │    33 │
│ 2022-03-16 15:04:00 │    33 │
│ 2022-03-16 15:04:00 │    33 │
│ 2022-03-16 15:04:00 │    33 │
│ 2022-03-16 15:05:00 │    33 │
│ 2022-03-16 15:05:00 │    33 │
│ 2022-03-16 15:06:00 │    33 │
│ 2022-03-16 15:07:00 │    33 │
│ 2022-03-16 15:08:00 │    33 │
│ 2022-03-16 15:09:00 │    33 │
│ 2022-03-16 15:10:00 │    33 │
│ 2022-03-16 15:11:00 │    33 │
└─────────────────────┴───────┘
```

**Expected behavior**

Remove the `WHERE` condition and we get:

```
┌────────TimeReceived─┬─SrcAS─┐
│ 2022-03-16 15:04:00 │ 15169 │
│ 2022-03-16 15:04:00 │ 12322 │
│ 2022-03-16 15:04:00 │ 60068 │
│ 2022-03-16 15:04:00 │ 15169 │
│ 2022-03-16 15:05:00 │ 15169 │
│ 2022-03-16 15:05:00 │ 15169 │
│ 2022-03-16 15:06:00 │ 15169 │
│ 2022-03-16 15:07:00 │ 15169 │
│ 2022-03-16 15:08:00 │ 15169 │
│ 2022-03-16 15:09:00 │ 15169 │
│ 2022-03-16 15:10:00 │ 15169 │
│ 2022-03-16 15:11:00 │ 15169 │
└─────────────────────┴───────┘
```

This would be the expected result.

Removing the subquery also works:

```sql
SELECT toStartOfInterval(toDateTime(TimeReceived), INTERVAL 20 second) AS TimeReceived,
       if(SrcAS in (15169, 12322) , SrcAS, 33) AS SrcAS
FROM flows
WHERE 2 == 2;
```

This gives:

```
┌────────TimeReceived─┬─SrcAS─┐
│ 2022-03-16 15:04:00 │ 15169 │
│ 2022-03-16 15:04:00 │ 12322 │
│ 2022-03-16 15:04:00 │    33 │
│ 2022-03-16 15:04:00 │ 15169 │
│ 2022-03-16 15:05:00 │ 15169 │
│ 2022-03-16 15:05:00 │ 15169 │
│ 2022-03-16 15:06:00 │ 15169 │
│ 2022-03-16 15:07:00 │ 15169 │
│ 2022-03-16 15:08:00 │ 15169 │
│ 2022-03-16 15:09:00 │ 15169 │
│ 2022-03-16 15:10:00 │ 15169 │
│ 2022-03-16 15:11:00 │ 15169 │
└─────────────────────┴───────┘
```

### PR DESCRIPTION
**Pull request by amosbird**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix projection analysis which might lead to wrong query result when IN subquery is used. This fixes https://github.com/ClickHouse/ClickHouse/issues/35336

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -273,14 +273,16 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
     PreparedSets prepared_sets_)
     /// NOTE: the query almost always should be cloned because it will be modified during analysis.
     : IInterpreterUnionOrSelectQuery(options_.modify_inplace ? query_ptr_ : query_ptr_->clone(), context_, options_)
     , storage(storage_)
     , input_pipe(std::move(input_pipe_))
     , log(&Poco::Logger::get("InterpreterSelectQuery"))
     , metadata_snapshot(metadata_snapshot_)
-    , prepared_sets(std::move(prepared_sets_))
+    , subquery_for_sets(std::make_shared<SubqueriesForSets>(std::move(subquery_for_sets_)))
+    , prepared_sets(prepared_sets_)
```
Comment (novikd) [inline src/Interpreters/InterpreterSelectQuery.cpp:285]: Why did you remove `std::move`?

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -551,8 +550,8 @@ InterpreterSelectQuery::InterpreterSelectQuery(
         LOG_TRACE(log, "Running 'analyze' second time");
 
         /// Reuse already built sets for multiple passes of analysis
-        subquery_for_sets = std::move(query_analyzer->getSubqueriesForSets());
-        prepared_sets = query_info.sets.empty() ? query_analyzer->getPreparedSets() : query_info.sets;
+        subquery_for_sets = std::make_shared<SubqueriesForSets>(std::move(query_analyzer->getSubqueriesForSets()));
```
Comment (novikd) [inline src/Interpreters/InterpreterSelectQuery.cpp:553]: `subquery_for_sets` is always initialized, so I'd prefer to use here move-assignment without new memory allocation.

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 3]
```diff
@@ -4848,9 +4848,12 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
         query_ptr,
         query_context,
         query_options,
-        /* prepared_sets_= */ query_info.sets);
+        std::move(*query_info.subquery_for_sets),
+        query_info.sets);
     const auto & analysis_result = select.getAnalysisResult();
+
     query_info.sets = select.getQueryAnalyzer()->getPreparedSets();
+    query_info.subquery_for_sets = std::make_shared<SubqueriesForSets>(std::move(select.getQueryAnalyzer()->getSubqueriesForSets()));
```
Comment (novikd) [inline src/Storages/MergeTree/MergeTreeData.cpp:4856]: The same applied here

File: src/Interpreters/InterpreterSelectQuery.h [hunk 4]
```diff
@@ -205,6 +208,7 @@ class InterpreterSelectQuery : public IInterpreterUnionOrSelectQuery
     StorageSnapshotPtr storage_snapshot;
 
     /// Reuse already built sets for multiple passes of analysis, possibly across interpreters.
+    std::shared_ptr<SubqueriesForSets> subquery_for_sets;
```
[the remark this answers is not in the document]
Comment (amosbird) [reply src/Interpreters/InterpreterSelectQuery.h:211]: The reason to use `std::shared_ptr` is somehow hacky: to make it copyable.

[the remark this answers is not in the document]
Comment (novikd) [reply src/Interpreters/InterpreterSelectQuery.h:212]: I understand but I didn't see where it's used for the creation of the copy.

[the remark this answers is not in the document]
Comment (amosbird) [reply src/Interpreters/InterpreterSelectQuery.h]: It's because we need `SelectQueryInfo` to be copyable.  

https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageMerge.cpp#L218

I'll use a different implementation.

**Review by novikd — APPROVED** [review]

**Comment by novikd** [comment]
@Mergifyio update
